src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala

```diff
@@ -0,0 +1,169 @@
+package com.comcast.xfinity.sirius.uberstore.data
+
+import java.io.{BufferedInputStream, DataInputStream, EOFException, InputStream, RandomAccessFile}
+import java.nio.ByteBuffer
+import java.nio.channels.FileChannel
+
+import scala.annotation.tailrec
+
+/**
+  * Trait providing abstraction of low level read operations to an Uberstore
+  */
+trait UberDataFileReader extends AutoCloseable {
+  /**
+    * Returns the current offset within the Uberstore
+    */
+  def offset(): Long
+
+  /**
+    * Reads the next byte from the file
+    * @return the value of the next byte in the range 0 to 255, or -1 if EOF
+    */
+  def read(): Int
+
+  /**
+    * Reads the next 32-bit Integer from the file
+    * @return the 32-bit Integer
+    * @throws EOFException if EOF
+    */
+  def readInt(): Int
+
+  /**
+    * Reads the next 64-bit Long from the file
+    * @return the 64-bit Long
+    * @throws EOFException if EOF
+    */
+  def readLong(): Long
+
+  /**
+    * Populates the entire array with the next bytes from the file
+    * @param array to populate
+    * @throws EOFException if EOF
+    */
+  def readFully(array: Array[Byte]): Unit
+
+  /**
+    * Helper function that attempts to read the next 32-bit Integer from the file
+    * @return [[scala.Some]] if could successfully read the next 32-bit Integer, otherwise [[scala.None]]
+    * @throws EOFException if can only partially read the 32-bit Integer
+    */
+  def tryReadInt(): Option[Int] = {
+    val byte1 = this.read()
+    if (byte1 < 0) {
+      return None
+    }
+    val byte2 = this.read()
+    val byte3 = this.read()
+    val byte4 = this.read()
+    if (byte2 < 0 || byte3 < 0 || byte4 < 0) {
+      throw new EOFException()
+    }
+    Some((byte1 << 24) + (byte2 << 16) + (byte3 << 8) + (byte4 << 0))
+  }
+}
+
+object UberDataFileReader {
+  def apply(dataFileName: String, baseOffset: Long, useReadBuffer: Boolean, readBufferSize: Int): UberDataFileReader =
+    if (useReadBuffer) {
+      sequentialBuffered(dataFileName, baseOffset, readBufferSize)
+    } else {
+      randomAccess(dataFileName, baseOffset)
+    }
+
+  def sequentialBuffered(dataFileName: String, baseOffset:Long, readBufferSize: Int): UberDataFileReader = {
+    val randomAccessFile = new RandomAccessFile(dataFileName, "r")
+    randomAccessFile.seek(baseOffset)
+    val fileChannel = randomAccessFile.getChannel
+    val fileChannelInputStream = new FileChannelInputStream(fileChannel)
+    val inputStream = new BufferedInputStream(fileChannelInputStream, readBufferSize)
+    new BufferedUberDataFileReader(inputStream, baseOffset)
+  }
+
+  def randomAccess(dataFileName: String, baseOffset: Long): UberDataFileReader = {
+    val randomAccessFile = new RandomAccessFile(dataFileName, "r")
+    randomAccessFile.seek(baseOffset)
+    new RandomAccessUberDataFileReader(randomAccessFile)
+  }
+}
+
+private[uberstore] class RandomAccessUberDataFileReader(val randomAccessFile: RandomAccessFile) extends UberDataFileReader {
```

Comment:
I think it'd be worth moving `RandomAccessUberDataFileReader` and `BufferedUberDataFileReader` to separate files at this point.